parquet-column/src/main/java/org/apache/parquet/io/api/Binary.java

```diff
@@ -330,36 +330,55 @@ public static Binary fromByteArray(final byte[] value) {
   }
 
   private static class ByteBufferBackedBinary extends Binary {
-    private transient ByteBuffer value;
-    private transient byte[] cachedBytes;
+    private ByteBuffer value;
+    private byte[] cachedBytes;
+    private final int offset;
+    private final int length;
 
-    public ByteBufferBackedBinary(ByteBuffer value, boolean isBackingBytesReused) {
+    public ByteBufferBackedBinary(ByteBuffer value, int offset, int length, boolean isBackingBytesReused) {
       this.value = value;
+      this.offset = offset;
+      this.length = length;
       this.isBackingBytesReused = isBackingBytesReused;
     }
 
     @Override
     public String toStringUsingUTF8() {
-      return UTF8.decode(value).toString();
+      int limit = value.limit();
+      value.limit(offset+length);
+      int position = value.position();
+      value.position(offset);
+      // no corresponding interface to read a subset of a buffer, would have to slice it
+      // which creates another ByteBuffer object or do what is done here to adjust the
+      // limit/offset and set them back after
+      String ret = UTF8.decode(value).toString();
+      value.limit(limit);
+      value.position(position);
+      return ret;
     }
 
     @Override
     public int length() {
-      return value.remaining();
+      return length;
     }
 
     @Override
     public void writeTo(OutputStream out) throws IOException {
       // TODO: should not have to materialize those bytes
-      out.write(getBytesUnsafe());
+      out.write(getBytesUnsafe(), offset, length);
     }
 
     @Override
     public byte[] getBytes() {
```

Comment:
I think this is going to have to be a separate enhancement. While the read from disk has been moved to ByteBuffers, not all of the decoding and assembly code has been changed to read out of ByteBuffers, so in some cases we are going to have to take a hit and make a copy (or in the case of an on-heap buffer, we are actually just unwrapping the underlying byte[]) until there can be a complete code path for ByteBuffers all the way from the disk to the materialized records.